crates/lair_keystore_api/src/sodium_secretstream.rs

```diff
@@ -0,0 +1,301 @@
+//! libsodium secretstream Async reader / writer wrappers.
+
+use futures::future::{BoxFuture, FutureExt};
+use futures::stream::{BoxStream, Stream, StreamExt};
+use one_err::*;
+use parking_lot::Mutex;
+use sodoken::secretstream::xchacha20poly1305 as sss;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncWriteExt;
+
+use crate::LairResult;
+use std::future::Future;
+use std::sync::Arc;
+
+/// Throw errors on the streams if a single message is > 8 KiB
+const MAX_FRAME: usize = 1024 * 8; // 8 KiB
+
+/// Traits related to sodium_secretstreams. Unless you're writing a new
+/// stream implementation, you probably don't need these.
+pub mod traits {
+    use super::*;
+
+    /// SodiumSecretStream - Sender
+    pub trait AsS3Sender<T>: 'static + Send + Sync
+    where
+        T: 'static + serde::Serialize + Send,
+    {
+        /// Send data to the remote side of this connection.
+        fn send(&self, t: T) -> BoxFuture<'static, LairResult<()>>;
+
+        /// Get outgoing encryption context key
+        fn get_enc_ctx_key(&self) -> sodoken::BufReadSized<{ sss::KEYBYTES }>;
+
+        /// Get incoming decryption context key
+        fn get_dec_ctx_key(&self) -> sodoken::BufReadSized<{ sss::KEYBYTES }>;
+
+        /// Shutdown the channel.
+        fn shutdown(&self) -> BoxFuture<'static, LairResult<()>>;
```

Comment:
Would it make sense for all shutdown methods to take `self`?

Reply:
The trait-object-level shutdown is going to be wrapped in an Arc, so we wouldn't be able to call it if it took `self`... We could do that at the concrete wrapper struct-level... but strictly speaking, all that does is limit your options...